src/blog/fearless-structured-concurrency.vmd

```diff
@@ -0,0 +1,542 @@
+0 to the 3'th power is 0!
+1 to the 3'th power is 1!
+2 to the 3'th power is 8!
+3 to the 3'th power is 27!
+4 to the 3'th power is 64!
+```
+
+Look at that! There are two `Completed 4 iterations!` printed out. [# If this doesn't happen for you, try running it a few more times. It's rather fickle.] Why is that?
+
+
+It's because the line `numIterationsCompleted = numIterationsCompleted + 1;` has three steps:
+
+ * Load `numIterationsCompleted` into register A.
+ * Add 1 to register A.
+ * Store register A into `numIterationsCompleted`.
+
+
+If two threads are running in parallel, we might see this happen:
+
+ * 4th thread loads 3 from `numIterationsCompleted` into register A.
+ * 5th thread loads 3 from `numIterationsCompleted` into register B.
+ * 4th thread adds 1 to register A, it now contains 4.
+ * 5th thread adds 1 to register B, it now contains 4.
+ * 4th thread stores register A into `numIterationsCompleted`, it now contains 4.
+ * 5th thread stores register B into `numIterationsCompleted`, it now contains 4.
+
+
+The problem is that they didn't see each others add operations; they're both adding 1 to the old value 3, to get 4.
+
+If instead the 4th thread's store happened before the 5th thread's load, we'd have gotten the correct answer 5.
+
+
+This is a *data race* bugs. They are _very_ difficult to detect, because they depend on the scheduling of the threads.
+
+Luckily, we've figured out how to avoid these problems, with some "concurrency ground rules":
+
+ 1. Multiple threads can read the same data, if nobody can modify it.
+ 1. If a thread can read data that another thread can modify, the data must be wrapped in a mutex. [# More on mutexes below.]
+ 1. If data is only visible to one thread, it can access that data freely.
+
+
+We usually need proper fear, respect, and discipline to stick to these rules. But when a language enforces these rules, we have *fearless concurrency*.
+
+
+<slice/>
+
+
+# Fearless Concurrency
+
+Plenty of languages (Clojure, Pony, Erlang, Rust, Dart, some others [# See Renato Athaydes' [Fearless concurrency: how Clojure, Rust, Pony, Erlang and Dart let you achieve that.](https://sites.google.com/a/athaydes.com/renato-athaydes/posts/fearlessconcurrencyhowclojurerustponyerlanganddartletyouachievethat)]) offer fearless concurrency, but we'll use Pony and Rust as examples.
+
+
+## Message Passing with Pony
+
+Pony doesn't let us access data from the outside scope, like in our above C examples. Instead, we "send" data between threads.
+
+
+We can send data if *either* of these is true:
+
+ * There are no other references to the data (it has the `iso` permission). [# A "permission" is something that travels along with a pointer at compile time, which affects how you may use that pointer. For example, Rust's `&` vs `&mut`, C++'s `*` vs `const *`, Vale's `&` vs `*`.]
+ * It is deeply immutable (it has the `imm` permission).
+
+To send an object to another thread, it must have the `imm` or `iso` permission.
```

Comment:
Actor, again.
Also `tag` can be shared

Reply:
Reworded to avoid the need for mentioning tag